keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,1655 @@
+clarified that status will always be stored reliably and can be used as
+proposed in this KEP.
+
+Kube-scheduler
+and resource driver communicate by modifying that status. The status is also
+how Kubernetes tracks that a driver has allocated the resource and on
+which nodes the resource is available.
+
+This approach is an intentional simplification compared to the
+PersistentVolume/PersistentVolumeClaim model for volumes where the additional
+PV object was used to capture status. That model allowed operations like
+pre-provisioning volumes and then having Kubernetes bind those to claims that
+get created later. For resources, the resource driver can and must handle such
+pre-provisioning internally. Kubernetes wouldn't know how to match
+pre-provisioned resources against claims because it has no understanding about the
+parameters.
+
+Allocation of a resource happens either immediately when a ResourceClaim gets
+created (“immediate allocation”) or when when a Pod is getting scheduled which
+needs the resource (“delayed allocation”),
+depending on a flag in the ResourceClaim spec. Pods reference resource
+claims in a new PodSpec.Resources list. Each resource in that list
+then can be made available to one or more containers in that Pod.
+
+Immediate allocation is useful when allocating a resource is expensive (for
+example, programming an FPGA) and the resource therefore is meant to be used by
+multiple different Pods, either in parallel or one after the other. The
+downside is that Pod resource requirements cannot be considered when choosing
+where to allocate. If a Pod then cannot run on the node (or nodes, for
+network-attached resources) where the resource is available because resources
+like RAM or CPU are exhausted there, then the Pod also cannot run elsewhere.
+Delayed allocation solves this by integrating allocation with Pod scheduling.
+
+Allocation must be complete before a Pod is allowed to be scheduled onto a
+node. This avoids scenarios where a Pod is permanently assigned to a node which
+can't fit the pod because of the pod's other resource requirements.
+
+When a PodTemplateSpec in an app controller spec references a ResourceClaim by
+name, all Pods created by that controller also use that name and thus share the
+resources allocated for that ResourceClaim. This might be supported by the
+resource driver and can be useful, depending on the characteristics of the
+resource.
+
+But often, each Pod is meant to have exclusive access to its own ResourceClaim
+instance instead. To support such ephemeral resources without having to modify
+all controllers that create Pods, an entry in the new PodSpec.Resources list
+can also be a ResourceClaimTemplate. When a Pod gets created, such a template
+will be used to create a normal ResourceClaim with the Pod as owner, and then
+the normal allocation of the resource takes place.
+
+For immediate allocation, scheduling Pods is simple because the
```

Comment:
more detail on what happens with the allocation when the node is not healthy or is scaled out is needed here.  it would also be good to know the impact a resource claim that is pre-allocated has with a workload or node that uses specific taints/tolerations/labels.  it seems trivial that a resource claim could allocate on a node that a pod could never schedule due to other constraints?

Reply:
The way I see it, immediate allocation and separate ResourceClaims are useful for resources that can be made available on all nodes in a cluster. Cluster admins are in control here and can refuse to create ResourceClasses with immediate allocation when it doesn't make sense.

The next combination is delayed allocation with separate ResourceClaims. When a Pod cannot run with the way how those resources are allocated at the moment, the deallocation mechanism kicks in and resources are freed until (in the worst case) all of them are deallocated and an allocation+scheduling attempt starts from scratch.

Then we have delayed allocation with ephemeral inline ResourceClaims. Because each Pod starts with its own, new ResourceClaim instances, there is no prior allocations that may have to be undone.

The last combination, immediate allocation with ephemeral inline ResourceClaims, again only makes sense for resources that then will be available on all nodes. It might have some advantages in that case over delayed allocation because the scheduler will know that it can simply wait for the resource driver to finish the allocation, instead of having to request information and only then triggering the allocation.